subscription.go

```diff
@@ -543,95 +588,118 @@ func (sc *SubscriptionClient) Run() error {
 
 					if closeStatus != -1 && closeStatus < 3000 && closeStatus > 4999 {
 						sc.context.Log(fmt.Sprintf("%s. Retry connecting...", err), "client", GQLInternal)
-						if err = sc.Reset(); err != nil {
+						if err = sc.Run(); err != nil {
 							sc.errorChan <- err
 							return
 						}
 					}
 
+					if isClosedSubscriptionError(err) {
+						_ = sc.Close()
+						return
+					}
+
 					if sc.onError != nil {
 						if err = sc.onError(sc, err); err != nil {
 							// end the subscription if the callback return error
-							sc.Close()
+							_ = sc.Close()
 							return
 						}
 					}
 					continue
 				}
 
 				sub := sc.context.GetSubscription(message.ID)
-				go sc.protocol.OnMessage(sc.context, sub, message)
+				go sc.protocol.OnMessage(sc.context, *sub, message)
 			}
 		}
 	}()
 
-	for atomic.LoadInt64(&sc.isRunning) > 0 {
+	for sc.getClientStatus() == scStatusRunning {
 		select {
-		case <-sc.context.Done():
-			return nil
+		case <-ctx.Done():
+			if sc.context.GetSubscriptionsLength() == 0 {
+				sc.setClientStatus(scStatusClosing)
+			}
+			break
```

Comment:
I am afraid that this `break` statement does not return from `Run()`. As such, if subscription is closed, this loop will get stack to infinite loop. Instead a `return nil` will exit the `Run()` as intended after a closed/cancelled subscription.

Reply:
This is exposed by an integration test, where we continuously unsubscribe-subscribe a handler. This triggers the `no running subscription. exiting..` flow. When the next `subscribe` manages to add a subscription, the `Run` loop will get stack, because it never `breaks` out of the select statement.

So, I get what is the purpose of the `break` there, but it reveals this synchronisation issue.

Reply:
My impression is that `clientStatus` complicates things, taken that there is already the context.Context to signal a cancelled context.

I would suggest to redesign `Run` to depend on context.Context and errors for either retrying Run or exiting. My understanding is that you always want to exit the subscription.Run when closed or subscriptions are empty, right? If so, both can be similar, and avoid dealing with the length of subscriptions anywhere in the Run method.